Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,342 @@
+    return M(a); // a, b, and c are released here.
+}
+```
+
+Example 3b: Many qubits allocated with a separate `use` statement for each variable.
+
+```qsharp
+operation NestedBlocks(n : Int) : Result {
+    use a = Qubit();
+    use b = Qubit[2 * n + 1];
+    use c = Qubit[n];
+    // ...
+    return M(q); // a, b, and c are released here.
+}
+```
+
+Example 5b: A clean qubit and a borrowed qubit. This can only be expressed with multiple statements, not with tuple destructuring.
+
+```qsharp
+operation UsingAndBorrowing() : Result {
+    use a = Qubit();
+    borrow b = Qubit();
+    // ...
+    return M(a); // a and b are released here.
+}
+```
+
+Example 5b: A qubit allocated in new block and released before the end of the outer block.
+
+```qsharp
+operation DifferentLifetime() : Result {
+    mutable r = Zero;
+    {
+        use q = Qubit();
+        // Apply operations...
+        set r = M(q);
+        // q is released here.
+    }
+    // Classical post-processing...
+    return r;
+}
+```
+
+# Implementation
+
+The proposal could be implemented in one of two ways:
+
+1. As syntactic sugar for the existing block statements.
+   The new `use` and `borrow` statements are transformed in the AST into `using` and `borrowing` block statements, taking in the remaining statements of the block automatically.
+   The scope statement could be transformed into `if (true) { }`.
+   This would require no change to code generation.
+2. As new AST nodes, similar to `let` and `mutable` nodes for classical values.
+   A new scope statement AST node is added.
+   Code generation must release qubits at the end of the scope, but this is already analogous to reference counting for certain data types like arrays in QIR.
+
+Both options seem similar in terms of complexity and implementation timeline, but option 2 seems like the cleaner solution.
+Neither option has any performance differences compared to the existing implementation.
+
+## Timeline
+
+There are no dependencies on other proposals or libraries.
+The work needed for either option appears to be minimal, since it is similar to the existing implementation.
+
+# Further Considerations
+
+## Related Mechanisms
+
+This proposal relies on the existing qubit management functionality that is already a core feature of Q#.
+It does not change the behavior of qubit management; it only provides new syntax for existing functionality.
+
+## Impact on Existing Mechanisms
+
+The scope statement added in this proposal can also be used to limit the scope of non-qubit variable bindings declared with `let` and `mutable`.
+This may be useful for limiting the lifetime of large data structures like arrays or for otherwise communicating the intent that a variable should only be used in a specific section within a callable.
+
+## Anticipated Interactions with Future Modifications
+
+Because the proposed modification is similar to existing functionality, it is not expected to have any impact on future language development.
```

Comment:
Would it make sense to expand on how the current and proposed behavior would each look when combining with other potential new features like allocating qubits directly into UDTs like `LittleEndian` and `FixedPoint`? For example, would we want to support syntax like `use qs = LittleEndian(Qubit[10]);` or `let qs = LittleEndian(use Qubit[10]);`?

Similarly, how would this proposal look with apply-and-return expressions? I recall you mentioning that something like `use q = Qubit() <~ H;` could be a reasonable way to prepare a qubit in the |+⟩ state, for example.

Reply:
It's good to think about. I'm hoping it's mostly orthogonal to this proposal, since changes to qubit initializer expressions shouldn't need to affect the overall `use` statement. However, as you mentioned if we change initializers to put `use` on the right-hand side, like `let qs = LittleEndian(use Qubit[10]);` then that would affect this proposal.

As discussed in the other thread, I'm in favor of keeping `use` on the left, even if initializers are expanded to contain non-`Qubit` expressions like `LittleEndian(Qubit[10]);`. I think it'd be good to come to consensus on the positioning of `use` and then we could have a separate proposal for adding UDT wrapping and state preparation to qubit initializers?

Reply:
I totally agree with `use` on the left; my only point in raising was that `use` on the right has come up enough it's helpful to document as a considered and rejected alternative. In particular, that `use` on the right makes the state preparation / UDT discussion harder is a great argument in favor of `use` on the left.

Reply:
Agreed, this would be very good to look into. We should also examine the exact syntax of the initializers in general - Qubit[n] seems like the odd one out if we want to remove new array expressions (which I think is a good thing).